File: Week_02-哈希表-集合-树-图/589.n叉树的前序遍历.cpp
/*
 * @lc app=leetcode.cn id=589 lang=cpp
 *
 * [589] N叉树的前序遍历
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
#if 1
    vector<int> preorder(Node* root) {
        vector<int> res;
        stack<Node*> st;
        Node* rt;
        if(root == NULL) return res;
        st.push(root);
        while (!st.empty()) {
            rt = st.top();
            st.pop();
            res.push_back(rt->val);
            int width = rt->children.size();
            for (int i = width-1; i >=0; --i) {
                st.push(rt->children[i]);
            }
        }
        return res;
     }
#endif
#if 0
    //递归法
    vector<int> res;
    vecot<int> preorder(Node* root) {
        if(root) {
            res.push_back(root->val);
            for(int i = 0; i < root->children.size(); i++) {
                preorder(root->children[i]);
            }
        }
        return res;
    }
#endif

};
// @lc code=end

File: Week_02-哈希表-集合-树-图/589.n-叉树的前序遍历.py
#
# @lc app=leetcode.cn id=589 lang=python
#
# [589] N 叉树的前序遍历
#

# @lc code=start
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution(object):
    def preorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        '''
        #递归法
        ans = list()
        def dfs(root):
            if not root:
                return
            ans.append(root.val)
            for child in root.children: #遍历各子节点
                dfs(child)
        dfs(root)
        return ans
        '''
        #迭代法
        ans = list()
        stack = list()
        if not root:
            return ans
        stack.append(root)
        while stack:
            root = stack.pop()
            ans.append(root.val)
            stack.extend(root.children[::-1])  #children[::-1]可以倒序列举所有的child，最后一个加入的最左边的child，这样先弹出的也是左孩子
        return ans
        
        
# @lc code=end

